### PR TITLE
[#191] Refactor neural networks layers definitions

### LINKED ISSUE #191: Refactor neural networks layers definitions
Refactor neural layers' definitions in the constructors of most Pytorch models from `chain.iterable` into something more readable.

### PR DESCRIPTION
**Pull request by Michal-Kolomanski**
closes #191

### CODE REVIEW

File: recommender/engines/base/base_neural_network.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        layers = [
+            Linear(input_dim, layers_sizes[0]),
+            BatchNorm1d(layers_sizes[0]),
+            ReLU(),
+        ]
+        # The middle layers
+        layers += list(
+            chain.from_iterable(
+                [
+                    [Linear(n_size, next_n_size), BatchNorm1d(next_n_size), ReLU()]
```
Comment (wujuu) [inline recommender/engines/base/base_neural_network.py:36]: This is a good generalization although we should add the ability to either include or exclude BatchNorm (which should default to include) and also a way to choose the hidden activation function (which should default to ReLU)

**Review by wujuu — CHANGES_REQUESTED** [review]
Minor changes needed. Also the MLPHistoryEmbedder should inherit from BaseNeuralNetwork as well

File: recommender/engines/base/base_neural_network.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        layers += list(
+            chain.from_iterable(
+                [
+                    [
+                        Linear(n_size, next_n_size),
+                        BatchNorm1d(next_n_size),
+                        activation(),
+                    ]
+                    if inc_batchnorm
+                    else [Linear(n_size, next_n_size), activation()]
```
Comment (wujuu) [inline recommender/engines/base/base_neural_network.py:51]: This could be extracted to a private function `_create_layer` and then used here and also in the lines 33-39

Comment (wujuu) [reply recommender/engines/base/base_neural_network.py:51]: More clarity and DRY code

**Review by wujuu — CHANGES_REQUESTED** [review]

**Review by wujuu — APPROVED** [review]
